mycolor <- c(   
  "cluster0" = "#006400",            
  "cluster1" = "#1E90FF",           
  "cluster2" = "#800080",          
  "cluster3" = "#8B4513",          
  "cluster4" = "#008B8B",          
  "cluster5" = "#FF4500",           
  "cluster6" = "#DAA520",           
  "conjunctive-tissue" = "#778899", 
  "cortex-L0" = "#C71585",           
  "cortex-L1" = "#4682B4",        
  "cortex-L2" = "#BDB76B",
  "2nd-cortex" = "#990000",       
  "cortex-L3" = "#2E8B57",          
  "diamond" = "#D87093",            
  "stamen" = "#6495ED",              
  "pith" = "#4B0082",               
  "fruit-bundle" = "#BC8F8F",        
  "fruit-bundle-adjacent" = "#6A5ACD",
  "pedicel" = "#8B8B00",            
  "pedicel-bundle" = "#FF8C00",     
  "pedicel-bundle-1" = "#654321",    
  "seed" = "#9400D3",               
  "sepal" = "#2F4F4F",            
  "sepal-base" = "#20B2AA",       
  "stipule" = "#CD5C5C",          
  "wall" = "#9370DB",
  "epi-cortex"="#FFA07A"
)

#========================function======================================
# rename gene ID
sanitize_gene_names <- function(seurat_obj) {
  for (assay_name in names(seurat_obj@assays)) {
    assay <- seurat_obj@assays[[assay_name]]
    # 处理矩阵行名（counts/data/scale.data）
    for (slot_name in c("counts", "data", "scale.data")) {
      mat <- slot(assay, slot_name)
      if (!is.null(rownames(mat))) {
        rownames(mat) <- gsub("FvH4-", "FvH4_", rownames(mat))  # 特定前缀替换
        rownames(mat) <- gsub("-", "_", rownames(mat))         
        slot(assay, slot_name) <- mat
      }
    }

    if ("var.features" %in% slotNames(assay)) {
      assay@var.features <- gsub("FvH4-", "FvH4_", assay@var.features)
      assay@var.features <- gsub("-", "_", assay@var.features)
    }

    if (inherits(assay, "SCTAssay")) {
      for (model_name in names(assay@SCTModel.list)) {
        model <- assay@SCTModel.list[[model_name]]
        rownames(model@feature.attributes) <- gsub("FvH4-", "FvH4_", rownames(model@feature.attributes))
        rownames(model@feature.attributes) <- gsub("-", "_", rownames(model@feature.attributes))
        assay@SCTModel.list[[model_name]] <- model
      }
    }
    seurat_obj@assays[[assay_name]] <- assay
  }
  # 处理降维结果
  for (reduction_name in names(seurat_obj@reductions)) {
    reduc <- seurat_obj@reductions[[reduction_name]]
    if (!is.null(rownames(reduc@feature.loadings))) {
      rownames(reduc@feature.loadings) <- gsub("FvH4-", "FvH4_", rownames(reduc@feature.loadings))
      rownames(reduc@feature.loadings) <- gsub("-", "_", rownames(reduc@feature.loadings))
      seurat_obj@reductions[[reduction_name]] <- reduc
    }
  }
  return(seurat_obj)
}
#========Annotate scRNA-seq celltype based on marker genes======================

  #' Annotate the celltype for all clusters based on the marker genes and the input celltype signatures for scRNA-seq dataset.
  #'
  #' @docType methods
  #' @name RNAAnnotateCelltype
  #' @rdname RNAAnnotateCelltype
  #'
  #' @param RNA Seurat object of clustered scRNA-seq dataset, generated by \code{\link{RNARunSeurat}} function.
  #' @param genes Data frame of differential expressed genes, generated by \code{\link{RNARunSeurat}} function.
  #' @param signatures The \code{signatures} can be character or data frame. If \code{signatures} is character, 
  #' it should be "human.immune.CIBERSORT", "mouse.brain.ALLEN", "mouse.all.facs.TabulaMuris" or "mouse.all.droplet.TabulaMuris".
  #' If \code{signatures} is a data frame, is should be a data frame of celltype signatures, with first column is celltype, and second column is
  #' gene symbol. Check the immune signatures from CIBERSORT (Newman et al., Nature Method, 2015) in example for details. Default is "human.immune.CIBERSORT".
  #' @param min.score Minimum score required. For one cluster, if the score for all celltypes are less than 
  #' \code{min.score}, the cluster will be annotated as "Others". Default is 0.
  #' @param orig.ident Orignal identity. If given, will use the original identity to annotate the celltypes. Default is NULL.
  #'
  #' @author Chenfei Wang
  #'
  #' @return A RNA Seurat object with annotated celltypes.
  #'
  #'
  #' @examples
  #' data(pbmc.RNA)
  #' data(human.immune.CIBERSORT)
  #' pbmc.RNA.res <- RNARunSeurat(inputMat = pbmc.RNA, project = "PBMC.scRNA.Seurat", min.c = 10, min.g = 100, dims.use = 1:15)
  #' pbmc.RNA.res$RNA <- RNAAnnotateCelltype(pbmc.RNA.res$RNA, pbmc.RNA.res$genes, human.immune.CIBERSORT, min.score = 0.05)
  #' str(pbmc.RNA.res$RNA)
  #'
  #' @importFrom Seurat DimPlot Idents
  #' @importFrom ggplot2 ggsave
  #' @export
  
  RNAAnnotateCelltype <- function(RNA, genes, signatures = "human.immune.CIBERSORT", min.score = 0, orig.ident = NULL, outdir = "."){
    if(is.null(orig.ident)){
      if(class(signatures) == "character"){
        signatures_name = signatures
        data(list = signatures)
        signatures = get(signatures)
      } else {
        signatures_name = ""
      }
      celltypes <- as.character(unique(signatures[,1]))
      signature_list <- sapply(1:length(celltypes),function(x){
        return(toupper(as.character(signatures[which(signatures[,1]==celltypes[x]),2])))})
      names(signature_list) <- celltypes
      
      cluster_celltype_score = sapply(as.integer(unique(RNA@meta.data$seurat_clusters))-1, function(x){
        idx = genes$cluster==x
        avglogFC = genes$avg_logFC[idx]
        names(avglogFC) = toupper(genes$gene[idx])
        score_cluster = sapply(signature_list, function(y){
          score = sum(avglogFC[y], na.rm = TRUE) / log2(length(y))
          return(score)
        })
      })
      colnames(cluster_celltype_score) = as.character(as.integer(unique(RNA@meta.data$seurat_clusters))-1)
      
      cellscore_max = apply(cluster_celltype_score, 2, max, na.rm = TRUE)
      # cellscore_max_celltype = apply(cluster_celltype_score, 2, function(x){
      #     if (max(x) < min.score){
      #         return("Others")
      #     }else{
      #         return(rownames(cluster_celltype_score)[which.max(x)])
      #     }
      # })
      cellscore_max_celltype = apply(cluster_celltype_score, 2, function(x){
        x_max <- ifelse(all(is.na(x)), -Inf, max(x, na.rm = TRUE))
        if (x_max < min.score || x_max == -Inf){
          return("Others")
        } else {
          return(rownames(cluster_celltype_score)[which.max(x)])
        }
      })
      
      
      RNA@meta.data$assign.ident = as.character(as.integer(RNA@meta.data$seurat_clusters)-1)
      current.cluster.ids = as.character(as.integer(unique(RNA@meta.data$seurat_clusters))-1)
      new.cluster.ids = cellscore_max_celltype
      
      RNA@meta.data$assign.score = cellscore_max[RNA@meta.data$assign.ident]
      RNA@meta.data$assign.ident = plyr::mapvalues(x = RNA@meta.data$assign.ident,
                                                   from = current.cluster.ids, to = new.cluster.ids)
      if(signatures_name == "human.immune.CIBERSORT") {
        current.cluster.ids = sort(celltypes)
        new.cluster.ids = c("DC", "Mast", "CD4Tconv", "NK", "pDC", "CD8T", 
                            "CD8Tex", "Endothelial", "Eosinophils", "Fibroblasts", "Mono/Macro", "Mono/Macro", 
                            "Mono/Macro", "B", "Mono/Macro", "Myofibroblasts", "B", "CD4Tconv", 
                            "Neutrophils", "Plasma", "DC", "Mast", "CD4Tconv", "NK", 
                            "pDC", "CD4Tconv", "TMKI67", "Treg")
        RNA@meta.data$assign.ident = plyr::mapvalues(x = RNA@meta.data$assign.ident,
                                                     from = current.cluster.ids, to = new.cluster.ids)
      }
      p = DimPlot(object = RNA, label = TRUE, pt.size = 0.2, group.by = "assign.ident", label.size = 3, repel = T)
      ggsave(file.path(outdir, paste0(RNA@project.name, "_annotated.png")), p, width=6, height=4)}
    else{
      RNA$assign.ident <- orig.ident
      p = DimPlot(object = RNA, label = TRUE, pt.size = 0.2, group.by = "assign.ident", label.size = 3, repel = T)
      ggsave(file.path(outdir, paste0(RNA@project.name, "_original.png")), p,  width=5.5, height=4)}
    
    return(RNA)    
  }
 #=============================GO function=================================================== 
  get_GO_strawberry<-function(list, type){
    
    GO2geneID <- readMappings('topGO_strawberry.txt', sep= '\t', IDsep = ',')
    geneID2GO <- inverseList(GO2geneID)
    geneNames <- names(geneID2GO)
    geneList <- factor(as.integer(geneNames %in% list))
    names(geneList) <- geneNames
    GOdata <- new('topGOdata', ontology = type, allGenes = geneList,
                  annot = annFUN.gene2GO, gene2GO = geneID2GO)
    
    allGO <- usedGO(object = GOdata)
    resultFisher <- runTest(GOdata, algorithm = 'classic', statistic = 'fisher')
    gene <- geneData(resultFisher)
    t <- GenTable(GOdata,classic = resultFisher, topNodes = length(allGO))
    #adj <- p.adjust(t$classic, method = 'fdr')
    #ret_mat <- as.data.frame(cbind(t,adj))
    ret_mat <- as.data.frame(t)
    convert <- function(x){
      x <- as.character(x)
      if (startsWith(x, '<')){
        return(as.numeric(substr(x, 2, nchar(x))))
      } else{return(as.numeric(x))}
    }
    ret_mat$filter <- sapply(ret_mat$classic, function(x) convert(x))
    ret_mat <- ret_mat[ret_mat$filter <= 0.05,]
    #ret_mat <- cbind(ret_mat$GO.ID, ret_mat$Term, ret_mat$adj)
    #par(cex=0.05)
    #showSigOfNodes(GOdata, score(resultFisher), firstSigNodes = 3, useInfo = 'all')
    return (ret_mat[-ncol(ret_mat)])
  }
  #======================cluster===re-annotation=======================================================
  add_annotation_to_seurat <- function(seurat_obj, annotation_df, anno_colname = "annotation") {
    if (!all(c("V1", "V2") %in% colnames(annotation_df))) {
      stop("not match!")
    }
    colnames(annotation_df) <- c("barcode", "annotation")
    if (!any(grepl("-1$", annotation_df$barcode)) && any(grepl("-1$", Cells(seurat_obj)))) {
      annotation_df$barcode <- paste0(annotation_df$barcode, "-1")
    }
    seurat_obj[[anno_colname]] <- annotation_df$annotation[match(Cells(seurat_obj), annotation_df$barcode)]
    return(seurat_obj)
  }
  #cluster2cluster function=========================================
  get_cluster_markers <- function(seurat_obj, 
                                  sample_name, 
                                  group.by = "annotation",
                                  assay = "SCT", 
                                  slot = "data",
                                  min.pct = 0.25, 
                                  logfc.threshold = 0.25,
                                  only.pos = FALSE,
                                  return.thresh = 0.01,
                                  test.use = "wilcox") {
    cluster_ids <- sort(unique(as.character(seurat_obj[[group.by]][, 1])))
    result_list <- list()
    original_ident <- Idents(seurat_obj)
    # 设置 Idents 为指定 group
    seurat_obj$tmp_group <- seurat_obj[[group.by]][, 1]
    Idents(seurat_obj) <- "tmp_group"
    
    for (i in 1:(length(cluster_ids) - 1)) {
      for (j in (i + 1):length(cluster_ids)) {
        ident1 <- cluster_ids[i]
        ident2 <- cluster_ids[j]
        message("Comparing: ", ident1, " vs ", ident2, " ...")
        # 差异分析
        markers <- FindMarkers(
          object = seurat_obj,
          ident.1 = ident1,
          ident.2 = ident2,
          assay = assay,
          slot = slot,
          min.pct = min.pct,
          logfc.threshold = logfc.threshold,
          only.pos = only.pos,
          test.use = test.use,
          return.thresh = return.thresh
        )
        # 添加 cluster 信息
        markers$cluster1 <- ident1
        markers$cluster2 <- ident2
        key <- paste0(sample_name, "_", ident1, "_vs_", ident2)
        result_list[[key]] <- markers
      }
    }
    Idents(seurat_obj) <- original_ident
    return(result_list)
  }
  